File: web/src/sync/transfers.ts
import { ApiError, NetworkError } from '../api/client'
import type { SyncApi } from '../api/types'
import {
  cancelCapture,
  finishCapture,
  setFileState,
  storeDownloadedBlob,
} from '../commands/recordings'
import { now, putRow, recordingTx } from '../commands/write'
import { baseContentType, CHUNK_MS, isNotUploaded, type LocalFileState } from '../db/recordings'
import { getKeepOffline, getStorage, setStorage } from '../db/meta'
import { pendingFor } from '../db/outbox'
import type { CrosstuneDb } from '../db/schema'
import { liveTune } from '../db/tunes'
import { defaultLocks, heldCaptureIds } from './captureLock'
import { isAuthFailure } from './errors'

export const QUOTA_PROBLEM = 'urn:crosstune:quota-exceeded'

// Without a lock manager, a capture stops sending chunks as soon as the recorder does, so
// a gap this wide means the tab (or the recorder inside it) is gone, not just between chunks.
export const STALE_CAPTURE_MS = 6 * CHUNK_MS

// With a lock manager, a row not on the held list is never live, so a much shorter gap
// (covering the window between a chunk landing and the next one starting) is enough.
export const CAPTURE_LOCK_GRACE_MS = 2 * CHUNK_MS

/** A capture whose chunks stopped arriving a while ago: finish it from what it has so no audio is lost. */
export async function recoverInterruptedCaptures(
  db: CrosstuneDb,
  now: number = Date.now(),
  locks: LockManager | undefined = defaultLocks(),
): Promise<void> {
  const capturing = await db.recording_files.where('local_state').equals('capturing').toArray()
  const held = await heldCaptureIds(locks)
  const staleAfter = held ? CAPTURE_LOCK_GRACE_MS : STALE_CAPTURE_MS
  for (const file of capturing) {
    // A held lock means the recorder that owns this capture is still alive, in this tab
    // or another, regardless of how long it has been since its last chunk landed.
    if (held?.has(file.id)) continue
    const fresh = typeof file.last_chunk_at === 'number' && now - file.last_chunk_at < staleAfter
    if (fresh) continue
    const chunks = await db.recording_chunks.where('recording_id').equals(file.id).sortBy('idx')
    if (chunks.length === 0) {
      await cancelCapture(db, file.id)
      continue
    }
    const tune = file.tune_id ? await db.tunes.get(file.tune_id) : undefined
    await finishCapture(db, file.id, {
      tuneId: liveTune(tune)?.id ?? null,
      mime: chunks[0]!.blob.type || 'audio/mp4',
      // The recorder never reported a duration for a recording this abandoned, so the chunk
      // count is the only record of how long it ran.
      durationMs: file.local_duration_ms ?? chunks.length * CHUNK_MS,
      recordedAt: file.recorded_at ?? new Date().toISOString(),
    })
  }
}

// Doubles with each consecutive transient failure, capped so a long outage still retries hourly-ish.
const RETRY_BASE_MS = 30_000
const RETRY_MAX_MS = 30 * 60_000

/** How long a transfer that failed `attempts` times in a row waits before the next try. */
export function retryDelayMs(attempts: number): number {
  return Math.min(RETRY_MAX_MS, RETRY_BASE_MS * 2 ** attempts)
}

/** A transient failure goes back to captured for the next pass, but not before a backoff
 * that doubles with each consecutive miss, so a persistent error does not resend the same
 * blob on every pass. The row keeps the reason, so a recording that keeps waiting can say why. */
async function scheduleRetry(db: CrosstuneDb, id: string, cause: unknown): Promise<void> {
  const file = await db.recording_files.get(id)
  const attempts = file?.upload_attempts ?? 0
  const delay = retryDelayMs(attempts)
  await db.recording_files.update(id, {
    local_state: 'captured',
    error: cause instanceof Error ? cause.message : String(cause),
    upload_attempts: attempts + 1,
    next_attempt_at: Date.now() + delay,
  })
}

/** Leave the retry loop, win or lose: a state outside it (uploaded, blocked, failed) starts fresh. */
async function settleUpload(
  db: CrosstuneDb,
  id: string,
  state: LocalFileState,
  error: string | null = null,
): Promise<void> {
  await db.recording_files.update(id, {
    local_state: state,
    error,
    upload_attempts: 0,
    next_attempt_at: null,
  })
}

async function uploadOne(db: CrosstuneDb, api: SyncApi, id: string): Promise<void> {
  const file = await db.recording_files.get(id)
  if (!file?.blob) return
  const row = await db.recordings.get(id)
  // A tombstone that lands mid-pass is settled by the next pass's dropTombstonedFiles.
  if (!row || row.deleted_at) return
  // The row must exist on the server before it can be given a slot.
  if (await pendingFor(db, 'recordings', id)) return
  if (row.state !== 'pending_upload') {
    await settleUpload(db, id, 'uploaded')
    return
  }
  await setFileState(db, id, 'uploading')
  const contentType = baseContentType(file.mime ?? file.blob.type)

  let slot
  try {
    slot = await api.requestUploadSlot(id, { bytes: file.blob.size, content_type: contentType })
  } catch (error) {
    if (error instanceof ApiError && error.problemType === QUOTA_PROBLEM) {
      await settleUpload(db, id, 'blocked_quota', error.message)
      return
    }
    if (error instanceof ApiError && error.status === 409) {
      // Already past pending on the server, so a previous confirmation landed.
      await settleUpload(db, id, 'uploaded')
      return
    }
    if (error instanceof ApiError && error.status === 404) {
      // The server has no live row for this recording; pushing it again gives the slot request one.
      await requeueRecording(db, id)
      return
    }
    if (
      error instanceof ApiError &&
      error.status >= 400 &&
      error.status < 500 &&
      ![401, 403, 408, 429].includes(error.status)
    ) {
      // A refusal of the request itself (bad mime, too large) will never succeed
      // as-is; an auth, timeout, or rate-limit refusal might on retry.
      await settleUpload(db, id, 'failed_upload', error.message)
      return
    }
    // Everything else (5xx, 401/403/408/429, network and auth failures, a transfer error) is transient.
    await scheduleRetry(db, id, error)
    throw error
  }

  try {
    await api.putObject(slot.url, file.blob, contentType)
    await api.uploadFinished(id)
    await settleUpload(db, id, 'uploaded')
  } catch (error) {
    if (error instanceof ApiError && error.status === 409) {
      // The slot expired or the object never landed; the next pass requests a fresh one.
      await scheduleRetry(db, id, error)
      return
    }
    if (error instanceof ApiError && error.status === 413) {
      // The server deleted the object because it did not match the declared size.
      await settleUpload(db, id, 'failed_upload', error.message)
      return
    }
    await scheduleRetry(db, id, error)
    throw error
  }
}

/** Queue the recording row again with a fresh timestamp and put its file back in the upload queue. */
async function requeueRecording(db: CrosstuneDb, id: string): Promise<void> {
  await recordingTx(db, async () => {
    const row = await db.recordings.get(id)
    if (row && !row.deleted_at) await putRow(db, 'recordings', { ...row, updated_at: now() })
    await db.recording_files.update(id, {
      local_state: 'captured',
      error: null,
      upload_attempts: 0,
      next_attempt_at: null,
    })
  })
}

/** Remove a file row (and its chunks) whose recording was tombstoned elsewhere, and any
 * chunk left with no file row or one that has moved past capturing. A capture still in
 * progress is left alone: it has no server row to check against yet, and still needs its chunks.
 * A tombstoned recording that never uploaded is restored as unfiled instead: the server takes a
 * newer upsert over its tombstone, and this blob is the only copy of the audio. */
async function dropTombstonedFiles(db: CrosstuneDb): Promise<void> {
  await recordingTx(db, async () => {
    const files = await db.recording_files.filter((f) => f.local_state !== 'capturing').toArray()
    const rows = await db.recordings.bulkGet(files.map((f) => f.id))
    for (const [i, file] of files.entries()) {
      const row = rows[i]
      if (row && !row.deleted_at) continue
      if (row && file.blob && isNotUploaded(file)) {
        await putRow(db, 'recordings', {
          ...row,
          deleted_at: null,
          tune_id: null,
          updated_at: now(),
        })
        await db.recording_files.update(file.id, { local_state: 'captured', error: null })
        continue
      }
      await db.recording_files.delete(file.id)
      await db.recording_chunks.where('recording_id').equals(file.id).delete()
    }

    // Read the owners off the primary keys rather than a unique-direction index cursor,
    // which Safari's IndexedDB refuses to open ("Unable to open cursor").
    const chunkKeys = await db.recording_chunks.toCollection().primaryKeys()
    const chunkOwners = [...new Set(chunkKeys.map(([recordingId]) => recordingId))]
    const owners = await db.recording_files.bulkGet(chunkOwners)
    for (const [i, recordingId] of chunkOwners.entries()) {
      if (owners[i]?.local_state === 'capturing') continue
      await db.recording_chunks.where('recording_id').equals(recordingId).delete()
    }
  })
}

/**
 * Upload every captured blob whose row has reached the server. Blocked and stalled ones
 * are retried too. A row's own transient failure does not stop the rest of the pass; the
 * first one is returned (null when every row settled) so the caller can still fail the sync.
 */
export async function uploadPass(db: CrosstuneDb, api: SyncApi): Promise<unknown> {
  // Run before the quota skip below, so a blocked row whose recording was deleted
  // elsewhere is cleaned up instead of being skipped forever by the figures check.
  await dropTombstonedFiles(db)

  const waiting = await db.recording_files
    .where('local_state')
    .anyOf(['captured', 'blocked_quota', 'uploading'])
    .toArray()
  const storage = await getStorage(db)
  const nowMs = Date.now()
  let firstError: unknown = null
  for (const file of waiting) {
    // A pass is the only place a blocked row is reconsidered, so re-check the figures
    // rather than hammering the server with a slot request that will fail again.
    if (
      file.local_state === 'blocked_quota' &&
      storage &&
      storage.used_bytes + file.bytes > storage.quota_bytes
    ) {
      continue
    }
    if (file.next_attempt_at !== null && file.next_attempt_at > nowMs) continue
    try {
      await uploadOne(db, api, file.id)
    } catch (error) {
      // Only an auth failure stops the whole pass; every other failure, including one
      // that never reached the network, is this row's problem alone and must not
      // block pull or the rest of the queue.
      if (isAuthFailure(error)) throw error
      firstError ??= error
    }
  }
  return firstError
}

/** A stale 'downloading' row is a prior attempt that never finished, not a state to
 * restore as-is: treat it the same as a completed fetch that produced no blob. */
function restingState(state: LocalFileState): LocalFileState {
  return state === 'downloading' ? 'downloaded' : state
}

export async function downloadOne(db: CrosstuneDb, api: SyncApi, id: string): Promise<Blob | null> {
  const file = await db.recording_files.get(id)
  if (file?.blob) return file.blob
  const row = await db.recordings.get(id)
  if (!row || row.deleted_at || row.state !== 'ready') return null
  const previousState = file ? restingState(file.local_state) : null
  await setFileState(db, id, 'downloading')
  try {
    const signed = await api.downloadUrl(id)
    const blob = await api.getObject(signed.url)
    await storeDownloadedBlob(db, id, blob, row.playback_mime ?? blob.type)
    return blob
  } catch (error) {
    // A row that had no state before downloading gets none back; setFileState no-ops on it.
    if (previousState) await setFileState(db, id, previousState)
    throw error
  }
}

export interface DownloadRetries {
  isWaiting(id: string): boolean
  failed(id: string): void
  succeeded(id: string): void
}

/** When the download pass may next try each recording whose download failed, backing off as
 * uploads do so a lasting error is not re-sent after every sync. Kept in memory: a reload
 * tries each once more, and a Play tap fetches regardless. */
export function createDownloadRetries(now: () => number = Date.now): DownloadRetries {
  const waits = new Map<string, { attempts: number; until: number }>()
  return {
    isWaiting: (id) => {
      const wait = waits.get(id)
      return wait !== undefined && now() < wait.until
    },
    failed: (id) => {
      const attempts = waits.get(id)?.attempts ?? 0
      waits.set(id, { attempts: attempts + 1, until: now() + retryDelayMs(attempts) })
    },
    succeeded: (id) => {
      waits.delete(id)
    },
  }
}

/** When the device keeps recordings offline, fetch audio for every ready recording that is
 * not already here. `fetch` lets the caller share one in-flight download per recording with
 * any other path that fetches. */
export async function downloadPass(
  db: CrosstuneDb,
  api: SyncApi,
  fetch: (id: string) => Promise<Blob | null> = (id) => downloadOne(db, api, id),
  retries: DownloadRetries = createDownloadRetries(),
): Promise<void> {
  if (!(await getKeepOffline(db))) return
  const rows = await db.recordings.filter((r) => !r.deleted_at && r.state === 'ready').toArray()
  const files = await db.recording_files.bulkGet(rows.map((r) => r.id))
  for (const [i, row] of rows.entries()) {
    const file = files[i]
    if (file?.blob || retries.isWaiting(row.id)) continue
    try {
      await fetch(row.id)
      retries.succeeded(row.id)
    } catch (error) {
      // Offline and auth failures stop the whole pass; anything else is this row's
      // problem alone, so record it and let the rest of the kept-offline set still download.
      if (error instanceof NetworkError || isAuthFailure(error)) {
        throw error
      }
      retries.failed(row.id)
      if (file) {
        const message = error instanceof Error ? error.message : String(error)
        await setFileState(db, row.id, restingState(file.local_state), message)
      }
    }
  }
}

export async function refreshStorage(db: CrosstuneDb, api: SyncApi): Promise<void> {
  const me = await api.me()
  await setStorage(db, me.storage)
}

File: apple/CrosstuneKit/Sources/CrosstuneSync/Transfers.swift
import CrosstuneStore
import Foundation
import GRDB

/// The problem type the API answers a slot request with when the recording would pass the
/// account's quota.
let quotaProblem = "urn:crosstune:quota-exceeded"

/// The first backoff after a transient upload failure, doubling with each one after it.
let uploadRetryBase: Duration = .seconds(30)
/// The longest an upload waits between tries, so a long outage still retries twice an hour.
let uploadRetryMax: Duration = .seconds(30 * 60)

/// How long a file that failed to upload `attempts` times in a row waits before the next try.
func uploadBackoff(attempts: Int) -> Duration {
    // Capped before the shift so a long streak cannot overflow.
    let factor = 1 << min(attempts, 16)
    return min(uploadRetryMax, uploadRetryBase * factor)
}

/// The type without codec parameters, which is what the signed upload is pinned to.
func baseContentType(_ contentType: String?) -> String {
    let base = contentType?.split(separator: ";", maxSplits: 1).first?.trimmingCharacters(in: .whitespaces)
    guard let base, !base.isEmpty else { return "application/octet-stream" }
    return base.lowercased()
}

/// The name a downloaded recording's file takes in the audio folder. The extension lets the
/// player tell the container without sniffing it.
func downloadedFileName(_ recordingID: String, contentType: String?) -> String {
    let ext =
        switch baseContentType(contentType) {
        case "audio/mpeg": "mp3"
        case "audio/webm": "webm"
        case "audio/ogg": "ogg"
        case "audio/wav", "audio/x-wav": "wav"
        default: "m4a"
        }
    return "\(recordingID).\(ext)"
}

/// Why a file waiting to upload is stuck when its audio is no longer on this device.
let missingAudio = "The audio file is missing from this device."

/// Why a transfer failed, as a recording's row shows it.
func transferMessage(_ error: any Error) -> String {
    switch underlying(error) {
    case let refusal as APIStatusError: refusal.message
    case let failure as TransferError: failure.message
    case let other: other.localizedDescription
    }
}

/// One run of the transfer loop's work: uploads every captured file whose recording has reached
/// the server, and downloads what this device keeps offline. Also the single download a play
/// asks for.
///
/// Every request is preceded by `checkStopped`, so a pass on an engine that has stopped for a
/// sign-out never sends the next user's token on the previous user's behalf.
@MainActor
struct Transfers {
    let store: CrosstuneStore
    let api: any SyncAPI
    let checkStopped: @MainActor () throws -> Void
    var downloadRetries = DownloadRetries()

    // MARK: Upload

    /// Uploads every file waiting to go. A file's own transient failure does not stop the rest;
    /// the first one is returned, nil when every file settled, so the run can still fail. A stop
    /// or an auth failure ends the pass at once.
    func uploadPass() async throws -> (any Error)? {
        // Before the quota skip below, so a blocked file whose recording was deleted elsewhere is
        // cleaned up instead of being skipped forever by the figures check.
        try await dropTombstonedFiles()

        let waiting: [LocalFileState] = [.captured, .blockedQuota, .uploading]
        let files = try await store.read { db in
            try RecordingFile.filter(waiting.contains(RecordingFile.CodingKeys.localState)).fetchAll(db)
        }
        let storage = try await store.meta(.storage, as: StorageFigures.self)
        let now = Timestamp.now
        var firstError: (any Error)?
        for file in files {
            // A pass is the only place a blocked file is reconsidered, so the figures are checked
            // rather than asking for a slot the server will refuse again.
            if file.localState == .blockedQuota, let storage,
                Int64(storage.usedBytes) + (file.bytes ?? 0) > Int64(storage.quotaBytes)
            {
                continue
            }
            if let next = file.nextAttemptAt, next > now { continue }
            do {
                try await uploadOne(file.id)
            } catch {
                if error is RunStopped || isAuthFailure(error) { throw error }
                firstError = firstError ?? error
            }
        }
        return firstError
    }

    private func uploadOne(_ id: String) async throws {
        let found = try await store.read { db -> (RecordingFile, String, Recording, Bool)? in
            guard let file = try RecordingFile.fetchOne(db, key: id), let fileName = file.fileName,
                let row = try Recording.fetchOne(db, key: id)
            else { return nil }
            let queued =
                try OutboxEntry.filter(
                    OutboxEntry.CodingKeys.tableName == SyncTable.recordings.rawValue
                        && OutboxEntry.CodingKeys.rowID == id
                ).fetchCount(db) > 0
            return (file, fileName, row, queued)
        }
        // A tombstone that lands mid-pass is settled by the next pass's dropTombstonedFiles.
        guard let (file, fileName, row, queued) = found, row.deletedAt == nil else { return }
        // The server gives a slot only to a recording it has.
        if queued { return }
        guard row.state == "pending_upload" else {
            try await settleUpload(id, .uploaded)
            return
        }
        let url = store.audioFolder.appending(path: fileName)
        guard FileManager.default.fileExists(atPath: url.path(percentEncoded: false)) else {
            // Nothing to send, so retrying would only grow the try count.
            try await settleUpload(id, .failedUpload, error: missingAudio)
            return
        }
        try await setFileState(id, .uploading)
        let contentType = baseContentType(file.contentType)

        let slot: URL
        do {
            let bytes = try fileSize(url)
            try checkStopped()
            slot = try await api.requestUploadSlot(recordingID: id, bytes: bytes, contentType: contentType)
        } catch let stop as RunStopped {
            throw stop
        } catch {
            if let refusal = underlying(error) as? APIStatusError {
                if refusal.problemType == quotaProblem {
                    try await settleUpload(id, .blockedQuota, error: refusal.message)
                    return
                }
                switch refusal.status {
                case 409:
                    // Already past pending on the server, so an earlier confirmation landed.
                    try await settleUpload(id, .uploaded)
                    return
                case 404:
                    // The server has no live row for this recording; pushing it again gives the
                    // slot request one.
                    try await requeueRecording(id)
                    return
                case 400..<500 where ![401, 403, 408, 429].contains(refusal.status):
                    // A refusal of the request itself (bad type, too large) never succeeds as it
                    // is; an auth, timeout, or rate-limit refusal might on a retry.
                    try await settleUpload(id, .failedUpload, error: refusal.message)
                    return
                default: break
                }
            }
            try await scheduleRetry(id, after: error)
            throw error
        }

        do {
            try checkStopped()
            try await api.putObject(slot, file: url, contentType: contentType)
            try checkStopped()
            try await api.uploadFinished(recordingID: id)
            try await settleUpload(id, .uploaded)
        } catch let stop as RunStopped {
            throw stop
        } catch {
            switch (underlying(error) as? APIStatusError)?.status {
            case 409:
                // The slot expired or the file never landed; the next pass asks for a fresh one.
                try await scheduleRetry(id, after: error)
            case 413:
                // The server deleted the file because it did not match the size declared.
                try await settleUpload(id, .failedUpload, error: transferMessage(error))
            default:
                try await scheduleRetry(id, after: error)
                throw error
            }
        }
    }

    private func fileSize(_ url: URL) throws -> Int64 {
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path(percentEncoded: false))
        return (attributes[.size] as? NSNumber)?.int64Value ?? 0
    }

    /// A transient failure goes back to captured for a later pass, but not before a backoff that
    /// doubles with each failure in a row, so a lasting error does not resend the same file on
    /// every pass. The file keeps the reason, so a recording that keeps waiting can say why.
    private func scheduleRetry(_ id: String, after error: any Error) async throws {
        let message = transferMessage(error)
        try await store.write { writer in
            guard var file = try RecordingFile.fetchOne(writer.db, key: id) else { return }
            let now = Timestamp.now
            let delay = uploadBackoff(attempts: file.uploadAttempts)
            file.localState = .captured
            file.error = message
            file.uploadAttempts += 1
            file.nextAttemptAt = Timestamp(
                milliseconds: now.milliseconds + delay.components.seconds * 1000)
            file.updatedAt = now
            try file.update(writer.db)
        }
    }

    /// Leaves the retry loop, win or lose: a state outside it (uploaded, blocked, failed) starts
    /// its count fresh.
    private func settleUpload(_ id: String, _ state: LocalFileState, error: String? = nil) async throws {
        try await store.write { writer in
            guard var file = try RecordingFile.fetchOne(writer.db, key: id) else { return }
            file.localState = state
            file.error = error
            file.uploadAttempts = 0
            file.nextAttemptAt = nil
            file.updatedAt = .now
            try file.update(writer.db)
        }
    }

    /// Queues the recording row again with a fresh time and puts its file back in the upload
    /// queue.
    private func requeueRecording(_ id: String) async throws {
        try await store.write { writer in
            if let row = try Recording.fetchOne(writer.db, key: id), row.deletedAt == nil {
                try writer.put(row)
            }
            guard var file = try RecordingFile.fetchOne(writer.db, key: id) else { return }
            file.localState = .captured
            file.error = nil
            file.uploadAttempts = 0
            file.nextAttemptAt = nil
            file.updatedAt = .now
            try file.update(writer.db)
        }
    }

    /// Removes the file and file row of a recording tombstoned elsewhere, or one with no
    /// recording row at all. A capture still in progress is left alone: it has no recording row
    /// yet. A tombstoned recording whose audio never uploaded is restored unfiled instead, since
    /// the server takes a newer upsert over its tombstone and this file is the only copy. One
    /// whose audio is gone from disk has no copy to save, so its delete stands.
    func dropTombstonedFiles() async throws {
        let audioFolder = store.audioFolder
        try await store.writeDroppingAudio { writer in
            let files = try RecordingFile.filter(RecordingFile.CodingKeys.localState != LocalFileState.capturing)
                .fetchAll(writer.db)
            for file in files {
                let row = try Recording.fetchOne(writer.db, key: file.id)
                if let row, row.deletedAt == nil { continue }
                if var row, let fileName = file.fileName, file.localState.isNotUploaded,
                    FileManager.default.fileExists(
                        atPath: audioFolder.appending(path: fileName).path(percentEncoded: false))
                {
                    row.deletedAt = nil
                    row.tuneID = nil
                    try writer.put(row)
                    var restored = file
                    restored.localState = .captured
                    restored.error = nil
                    restored.updatedAt = .now
                    try restored.update(writer.db)
                    continue
                }
                try file.delete(writer.db)
            }
        }
    }

    // MARK: Download

    /// When this device keeps recordings offline, fetches the audio of every ready recording not
    /// already here. `fetch` shares one download per recording with a play that asks for it.
    func downloadPass(fetch: @MainActor (String) async throws -> URL?) async throws {
        guard try await store.meta(.keepOffline, as: Bool.self) == true else { return }
        let missing = try await store.read { db -> [(String, RecordingFile?)] in
            let rows = try Recording.filter(
                Recording.CodingKeys.deletedAt == nil && Recording.CodingKeys.state == "ready"
            ).fetchAll(db)
            let files = Dictionary(
                uniqueKeysWithValues: try RecordingFile.fetchAll(db, keys: rows.map(\.id)).map { ($0.id, $0) })
            return rows.compactMap { row in
                let file = files[row.id]
                return file?.fileName == nil ? (row.id, file) : nil
            }
        }
        for (id, file) in missing where !downloadRetries.isWaiting(id) {
            do {
                _ = try await fetch(id)
                downloadRetries.succeeded(id)
            } catch {
                // A stop, no connection, or a refused session ends the pass; anything else is this
                // recording's problem alone, so it is noted and the rest still download.
                if error is RunStopped || underlying(error) is URLError || isAuthFailure(error) { throw error }
                downloadRetries.failed(id)
                if let file {
                    try await setFileState(id, restingState(file.localState), error: transferMessage(error))
                }
            }
        }
    }

    /// The recording's audio on this device, fetched first when it is not here. Nil when there
    /// is nothing to fetch: the recording is gone or not ready.
    func downloadOne(_ id: String) async throws -> URL? {
        let (file, row) = try await store.read { db in
            (try RecordingFile.fetchOne(db, key: id), try Recording.fetchOne(db, key: id))
        }
        if let local = store.localAudio(file) { return local }
        guard let row, row.deletedAt == nil, row.state == "ready" else { return nil }
        // Before the first write, so a stopped engine leaves the store as it found it.
        try checkStopped()
        let previousState = file.map { restingState($0.localState) }
        try await setFileState(id, .downloading)
        let name = downloadedFileName(id, contentType: row.playbackMime)
        let destination = store.audioFolder.appending(path: name)
        do {
            try checkStopped()
            let signed = try await api.downloadURL(recordingID: id)
            try checkStopped()
            try await api.getObject(signed, to: destination)
            // The server keeps the copy, so a device backup would only duplicate it.
            var values = URLResourceValues()
            values.isExcludedFromBackup = true
            var excluded = destination
            try excluded.setResourceValues(values)
            let bytes = try fileSize(destination)
            try await store.write { writer in
                var stored =
                    try RecordingFile.fetchOne(writer.db, key: id) ?? RecordingFile(id: id, localState: .downloaded)
                stored.localState = .downloaded
                stored.fileName = name
                stored.contentType = row.playbackMime
                stored.bytes = bytes
                stored.error = nil
                stored.updatedAt = .now
                try stored.save(writer.db)
            }
            return destination
        } catch {
            try? FileManager.default.removeItem(at: destination)
            // A file with no row before downloading gets none back; setFileState passes it by.
            if let previousState { try? await setFileState(id, previousState) }
            throw error
        }
    }

    /// A stale `downloading` state is an earlier fetch that never finished, not a state to keep:
    /// it rests as a completed fetch that produced no file.
    private func restingState(_ state: LocalFileState) -> LocalFileState {
        state == .downloading ? .downloaded : state
    }

    private func setFileState(_ id: String, _ state: LocalFileState, error: String? = nil) async throws {
        try await store.write { writer in
            guard var file = try RecordingFile.fetchOne(writer.db, key: id) else { return }
            file.localState = state
            file.error = error
            file.updatedAt = .now
            try file.update(writer.db)
        }
    }
}

/// When the download pass may next try each recording whose download failed, doubling the wait
/// with each failure in a row as uploads do, so a lasting error is not re-sent after every sync.
/// Kept in memory: a relaunch tries each once more, and a play fetches regardless.
@MainActor
final class DownloadRetries {
    var now: () -> ContinuousClock.Instant = { .now }
    private var waits: [String: (attempts: Int, until: ContinuousClock.Instant)] = [:]

    func isWaiting(_ id: String) -> Bool {
        guard let wait = waits[id] else { return false }
        return now() < wait.until
    }

    func failed(_ id: String) {
        let attempts = waits[id]?.attempts ?? 0
        waits[id] = (attempts + 1, now().advanced(by: uploadBackoff(attempts: attempts)))
    }

    func succeeded(_ id: String) {
        waits[id] = nil
    }
}

extension CrosstuneStore {
    /// The file a recording's row names, when it holds finished audio that is still on disk.
    public func localAudio(_ file: RecordingFile?) -> URL? {
        guard let file, let name = file.fileName, file.localState != .capturing, file.localState != .downloading
        else { return nil }
        let url = audioFolder.appending(path: name)
        return FileManager.default.fileExists(atPath: url.path(percentEncoded: false)) ? url : nil
    }

    /// The recording's audio on this device, without fetching it: what plays with no sync engine.
    public func localAudio(recordingID: String) async -> URL? {
        let file = try? await read { db in try RecordingFile.fetchOne(db, key: recordingID) }
        return localAudio(file ?? nil)
    }
}
